File: src/lib/engineIpc.ts
// Typed wrappers around the engine IPC surface. These mirror the Rust
// types in cutty-engine (serde camelCase). The engine owns all timeline
// state — the frontend only sends these commands and renders the
// EngineSnapshot events that come back.

import { invoke } from "@tauri-apps/api/core";
import { listen, type UnlistenFn } from "@tauri-apps/api/event";

export interface ProjectSettings {
  width: number;
  height: number;
  fps: number;
}

export type MediaKind = "video" | "audio" | "image" | "gif";

export interface MediaRef {
  id: number;
  path: string;
  /** Seconds; 0 exactly on still images. */
  duration: number;
  hasVideo: boolean;
  hasAudio: boolean;
  hasAlpha: boolean;
  kind: MediaKind;
}

export type TrackKind = "video" | "audio" | "text";

/** Per-track toggle switches (mirrors the engine's TrackFlag). */
export type TrackFlag = "locked" | "muted" | "hidden";

export const BLEND_MODES = [
  "normal",
  "multiply",
  "screen",
  "overlay",
  "add",
] as const;
export type BlendMode = (typeof BLEND_MODES)[number];

export interface Transform {
  x: number;
  y: number;
  scale: number;
  rotation: number;
}

/** Transition bound to a clip's out cut (kind = shader registry id). */
export interface Transition {
  kind: string;
  duration: number;
}

export type TextAlign = "left" | "center" | "right";

/** Visual style of a text clip. Pixel quantities are project pixels at
 * transform scale 1; colors are #RRGGBB or #RRGGBBAA. */
export interface TextStyle {
  fontFamily: string;
  /** 100–900 (400 regular, 700 bold). */
  weight: number;
  fontSize: number;
  fill: string;
  strokeColor: string;
  strokeWidth: number;
  shadowColor: string;
  shadowOffsetX: number;
  shadowOffsetY: number;
  /** 0..1; 0 disables the shadow. */
  shadowAlpha: number;
  align: TextAlign;
}

/** Styled text payload of a clip on a text track. */
export interface TextSpec {
  content: string;
  style: TextStyle;
}

/** Easing of the segment from a keyframe to the next. */
export type Easing = "linear" | "easeIn" | "easeOut" | "easeInOut";

/** One keyframe on a clip property lane. `t` is clip-relative seconds
 * (from the clip's timelineIn) — automation moves with the clip. */
export interface Keyframe {
  t: number;
  value: number;
  easing: Easing;
}

/** Keyframable clip property (volume now; transform/opacity in Phase 3).
 * Volume keyframes are a gain *multiplier* on top of the clip's static
 * `volume`. */
export type KeyframeProp = "volume";

/** Which clip edge a fade handle drags. */
export type FadeSide = "in" | "out";

export interface Clip {
  id: number;
  /** Absent exactly on text clips (which carry `text` instead). */
  mediaId?: number;
  timelineIn: number;
  timelineOut: number;
  sourceIn: number;
  sourceOut: number;
  transform: Transform;
  opacity: number;
  blendMode: BlendMode;
  speed: number;
  volume: number;
  /** Transition into the next clip on the same track (absent = none). */
  transitionOut?: Transition | null;
  /** Present exactly on text-track clips. */
  text?: TextSpec | null;
  /** Keyframe lanes by property; absent = no animation. Lanes are
   * sorted by `t` and never empty (engine invariants). */
  keyframes?: Partial<Record<KeyframeProp, Keyframe[]>>;
}

export interface Track {
  id: number;
  kind: TrackKind;
  name: string;
  /** Rejects edits (the engine enforces; the UI also pre-checks). */
  locked: boolean;
  /** Audio silenced (audio tracks and video clips' embedded audio). */
  muted: boolean;
  /** Video track excluded from the composite (preview and export). */
  hidden: boolean;
  clips: Clip[];
}

export interface Project {
  settings: ProjectSettings;
  media: MediaRef[];
  tracks: Track[];
}

/** A transition resolved to its effective span (engine-computed — the
 * chip renders exactly here; all clamping is engine-side). */
export interface TransitionSpan {
  trackId: number;
  fromClipId: number;
  toClipId: number;
  kind: string;
  /** The cut instant the transition is centered on. */
  cut: number;
  /** Effective span, seconds. */
  start: number;
  end: number;
  /** Stored (requested) duration, seconds. */
  requested: number;
  /** Longest duration this cut currently supports (drag clamp). */
  maxDuration: number;
}

export interface EngineSnapshot {
  project: Project;
  transitions: TransitionSpan[];
  undoDepth: number;
  redoDepth: number;
  transactionActive: boolean;
}

export type TrimEdge = "start" | "end";

export interface SnappedMove {
  timelineIn: number;
  snapPoint: number | null;
}

export const ENGINE_STATE_EVENT = "engine://project";

/** Subscribe to engine state snapshots (fires after every mutation). */
export function onEngineState(
  handler: (snapshot: EngineSnapshot) => void,
): Promise<UnlistenFn> {
  return listen<EngineSnapshot>(ENGINE_STATE_EVENT, (e) => handler(e.payload));
}

export function engineGetState(): Promise<EngineSnapshot> {
  return invoke<EngineSnapshot>("engine_get_state");
}

export function engineAddMedia(
  path: string,
  duration: number,
  hasVideo: boolean,
  hasAudio: boolean,
  hasAlpha: boolean,
  kind: MediaKind,
): Promise<number> {
  return invoke<number>("engine_add_media", {
    path,
    duration,
    hasVideo,
    hasAudio,
    hasAlpha,
    kind,
  });
}

/** Remove a media file and every clip referencing it (one undo step). */
export function engineRemoveMedia(mediaId: number): Promise<void> {
  return invoke("engine_remove_media", { mediaId });
}

export function engineAddClip(
  trackId: number,
  mediaId: number,
  timelineIn: number,
  sourceIn: number,
  sourceOut: number,
): Promise<number> {
  return invoke<number>("engine_add_clip", {
    trackId,
    mediaId,
    timelineIn,
    sourceIn,
    sourceOut,
  });
}

/** Place a new text clip. `trackId: null` = CapCut placement: topmost
 * text lane with room, else a new lane on top (one undo step). Returns
 * the new clip's id. */
export function engineAddTextClip(
  timelineIn: number,
  duration: number,
  text: TextSpec,
  transform?: Transform,
  trackId?: number,
): Promise<number> {
  return invoke<number>("engine_add_text_clip", {
    timelineIn,
    duration,
    text,
    transform: transform ?? null,
    trackId: trackId ?? null,
  });
}

/** Replace a text clip's content and/or style (equal payloads no-op). */
export function engineSetClipText(
  clipId: number,
  text: TextSpec,
): Promise<void> {
  return invoke("engine_set_clip_text", { clipId, text });
}

/** Distinct system font families, sorted. First call loads the font
 * database backend-side; cache the promise. */
export function textFontFamilies(): Promise<string[]> {
  return invoke<string[]>("text_font_families");
}

let fontFamiliesCache: Promise<string[]> | null = null;

/** [`textFontFamilies`], fetched once and shared. */
export function cachedFontFamilies(): Promise<string[]> {
  fontFamiliesCache ??= textFontFamilies();
  return fontFamiliesCache;
}

/** Measure a text block in project pixels at scale 1 (gizmo box size). */
export function textMeasure(text: TextSpec): Promise<[number, number]> {
  return invoke<[number, number]>("text_measure", { text });
}

export function engineMoveClip(
  clipId: number,
  timelineIn: number,
): Promise<void> {
  return invoke("engine_move_clip", { clipId, timelineIn });
}

/** Returns the clamped edge time the engine actually applied. */
export function engineTrimClip(
  clipId: number,
  edge: TrimEdge,
  to: number,
): Promise<number> {
  return invoke<number>("engine_trim_clip", { clipId, edge, to });
}

/** Returns the new right half's clip id. */
export function engineSplitClip(clipId: number, at: number): Promise<number> {
  return invoke<number>("engine_split_clip", { clipId, at });
}

export function engineDeleteClip(clipId: number): Promise<void> {
  return invoke("engine_delete_clip", { clipId });
}

export function engineRippleDelete(clipId: number): Promise<void> {
  return invoke("engine_ripple_delete", { clipId });
}

/** Set a clip's audio gain (linear; 1.0 = unity, 0.0 = silent). */
export function engineSetClipVolume(
  clipId: number,
  volume: number,
): Promise<void> {
  return invoke("engine_set_clip_volume", { clipId, volume });
}

/** Add a keyframe (or replace the one at that time). `t` is
 * clip-relative seconds; returns where it landed after clamping. */
export function engineAddKeyframe(
  clipId: number,
  prop: KeyframeProp,
  t: number,
  value: number,
  easing?: Easing,
): Promise<number> {
  return invoke<number>("engine_add_keyframe", {
    clipId,
    prop,
    t,
    value,
    easing: easing ?? null,
  });
}

/** Move the keyframe at `fromT` to `toT` with a new value (a dot drag
 * moves both axes). Returns the applied time after neighbor clamping —
 * feed it back as the next step's `fromT`. */
export function engineMoveKeyframe(
  clipId: number,
  prop: KeyframeProp,
  fromT: number,
  toT: number,
  value: number,
): Promise<number> {
  return invoke<number>("engine_move_keyframe", {
    clipId,
    prop,
    fromT,
    toT,
    value,
  });
}

/** Remove the keyframe at clip-relative time `t`. */
export function engineRemoveKeyframe(
  clipId: number,
  prop: KeyframeProp,
  t: number,
): Promise<void> {
  return invoke("engine_remove_keyframe", { clipId, prop, t });
}

/** Set a fade-in/out duration in seconds (0 removes it) — sugar over
 * the volume keyframe lane. Returns the applied (clamped) duration. */
export function engineSetClipFade(
  clipId: number,
  side: FadeSide,
  duration: number,
): Promise<number> {
  return invoke<number>("engine_set_clip_fade", { clipId, side, duration });
}

/** Extract a video clip's audio onto an audio track (one undo step);
 * the video clip's own volume drops to 0. Returns the new clip's id. */
export function engineExtractAudio(clipId: number): Promise<number> {
  return invoke<number>("engine_extract_audio", { clipId });
}

/** Set a clip's 2D placement (position/scale/rotation). */
export function engineSetClipTransform(
  clipId: number,
  transform: Transform,
): Promise<void> {
  return invoke("engine_set_clip_transform", { clipId, transform });
}

/** Set a clip's opacity (0.0 transparent .. 1.0 opaque). */
export function engineSetClipOpacity(
  clipId: number,
  opacity: number,
): Promise<void> {
  return invoke("engine_set_clip_opacity", { clipId, opacity });
}

/** Set how a clip blends with the layers below it. */
export function engineSetClipBlendMode(
  clipId: number,
  mode: BlendMode,
): Promise<void> {
  return invoke("engine_set_clip_blend_mode", { clipId, mode });
}

/** Move a clip onto another track (and position) in one step. */
export function engineMoveClipToTrack(
  clipId: number,
  trackId: number,
  timelineIn: number,
): Promise<void> {
  return invoke("engine_move_clip_to_track", { clipId, trackId, timelineIn });
}

/** Insert a new empty track at panel index (0 = top; engine clamps).
 * Returns the new track's id. */
export function engineAddTrack(
  kind: TrackKind,
  index: number,
): Promise<number> {
  return invoke<number>("engine_add_track", { kind, index });
}

/** Remove a track with all its clips (one undo step). */
export function engineRemoveTrack(trackId: number): Promise<void> {
  return invoke("engine_remove_track", { trackId });
}

/** Move a track to another panel position (restacks the composite). */
export function engineMoveTrack(trackId: number, to: number): Promise<void> {
  return invoke("engine_move_track", { trackId, to });
}

/** Flip a per-track flag (lock / mute / hide). */
export function engineSetTrackFlag(
  trackId: number,
  flag: TrackFlag,
  value: boolean,
): Promise<void> {
  return invoke("engine_set_track_flag", { trackId, flag, value });
}

export function engineUndo(): Promise<boolean> {
  return invoke<boolean>("engine_undo");
}

export function engineRedo(): Promise<boolean> {
  return invoke<boolean>("engine_redo");
}

export function engineBeginTransaction(): Promise<void> {
  return invoke("engine_begin_transaction");
}

export function engineCommitTransaction(): Promise<void> {
  return invoke("engine_commit_transaction");
}

export function engineRollbackTransaction(): Promise<void> {
  return invoke("engine_rollback_transaction");
}

/** One transition shader available in the library panel. */
export interface TransitionDef {
  id: string;
  label: string;
  defaultDuration: number;
}

/** The transition catalog (static per build; cache the result). */
export function transitionList(): Promise<TransitionDef[]> {
  return invoke<TransitionDef[]>("transition_list");
}

let transitionCatalogCache: Promise<TransitionDef[]> | null = null;

/** [`transitionList`], fetched once and shared. */
export function cachedTransitionList(): Promise<TransitionDef[]> {
  transitionCatalogCache ??= transitionList();
  return transitionCatalogCache;
}

/** Set, replace, or remove (null) the transition at a clip's out cut.
 * Returns the stored duration after engine-side clamping. */
export function engineSetTransition(
  clipId: number,
  transition: Transition | null,
): Promise<number | null> {
  return invoke<number | null>("engine_set_transition", {
    clipId,
    transition,
  });
}

/** Snap a time to clip edges / playhead; null when nothing is in range. */
export function engineSnapTime(
  t: number,
  threshold: number,
  playhead: number | null,
  exclude: number[],
): Promise<number | null> {
  return invoke<number | null>("engine_snap_time", {
    t,
    threshold,
    playhead,
    exclude,
  });
}

/** Snap a clip-move gesture (both clip edges compete for candidates). */
export function engineSnapClipMove(
  clipId: number,
  desiredIn: number,
  threshold: number,
  playhead: number | null,
): Promise<SnappedMove> {
  return invoke<SnappedMove>("engine_snap_clip_move", {
    clipId,
    desiredIn,
    threshold,
    playhead,
  });
}

File: src-tauri/src/engine_ipc.rs
//! Thin IPC wrappers around `cutty-engine`. No editing logic lives here:
//! commands translate ids/enums, forward to the engine, map errors to
//! strings, and emit a full-state snapshot event after every mutation.

use std::sync::{Arc, Mutex};

use cutty_engine::{
    transition_spans, BlendMode, ClipId, Easing, Engine, FadeSide, KeyframeProp, MediaId, Project,
    ProjectSettings, SnappedMove, TextSpec, TrackFlag, TrackId, TrackKind, Transform, Transition,
    TrimEdge,
};
use tauri::{AppHandle, Emitter, State};

/// Event carrying the full engine state after every committed or transient
/// change. Projects are a few KB of JSON, so full snapshots keep the
/// frontend trivially consistent (matching `EngineEvent::ProjectChanged`).
pub const ENGINE_STATE_EVENT: &str = "engine://project";

/// Managed state: the single engine instance behind a mutex. Engine
/// operations are microseconds on Phase 1 projects, so commands run
/// synchronously on the IPC thread. `Arc` so the autosave worker can hold
/// a reference too.
pub struct EngineHandle(pub Arc<Mutex<Engine>>);

impl Default for EngineHandle {
    fn default() -> Self {
        Self(Arc::new(Mutex::new(
            Engine::new(ProjectSettings::default()),
        )))
    }
}

/// A transition resolved to its effective on-timeline span (mirrors
/// `cutty_engine::TransitionSpan`). The UI renders chips exactly where
/// these say — clamping and centering stay engine-side.
#[derive(Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct TransitionSpanWire {
    pub track_id: u64,
    pub from_clip_id: u64,
    pub to_clip_id: u64,
    pub kind: String,
    pub cut: f64,
    pub start: f64,
    pub end: f64,
    /// Stored (requested) duration, seconds.
    pub requested: f64,
    /// The longest duration this cut currently supports.
    pub max_duration: f64,
}

/// Full engine state as sent to the frontend (also the return value of
/// [`engine_get_state`] for the initial fetch).
#[derive(Clone, serde::Serialize)]
#[serde(rename_all = "camelCase")]
pub struct EngineSnapshot {
    pub project: Project,
    /// Resolved transition spans for the project above (derived state —
    /// recomputed with every snapshot).
    pub transitions: Vec<TransitionSpanWire>,
    pub undo_depth: usize,
    pub redo_depth: usize,
    pub transaction_active: bool,
}

/// Wire form of [`TrimEdge`].
#[derive(Clone, Copy, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub enum TrimEdgeArg {
    Start,
    End,
}

impl From<TrimEdgeArg> for TrimEdge {
    fn from(edge: TrimEdgeArg) -> Self {
        match edge {
            TrimEdgeArg::Start => TrimEdge::Start,
            TrimEdgeArg::End => TrimEdge::End,
        }
    }
}

fn snapshot(engine: &mut Engine) -> EngineSnapshot {
    // The snapshot below supersedes any pending per-command events.
    engine.drain_events();
    let transitions = transition_spans(engine.project())
        .into_iter()
        .map(|s| TransitionSpanWire {
            track_id: s.track_id.0,
            from_clip_id: s.from_clip.0,
            to_clip_id: s.to_clip.0,
            kind: s.kind,
            cut: s.cut,
            start: s.start,
            end: s.end,
            requested: s.requested,
            max_duration: s.max_duration,
        })
        .collect();
    EngineSnapshot {
        project: engine.project().clone(),
        transitions,
        undo_depth: engine.undo_depth(),
        redo_depth: engine.redo_depth(),
        transaction_active: engine.transaction_active(),
    }
}

/// Emit the full snapshot and fan the change out to every mirror: the
/// playback engine and the persistence layer (dirty meta + autosave).
/// Also used by project load/new/restore after swapping the engine.
pub(crate) fn emit_state(app: &AppHandle, engine: &mut Engine) {
    let _ = app.emit(ENGINE_STATE_EVENT, snapshot(engine));
    // The playback engine mirrors every project change (its scrub/pump
    // paths resolve against the newest snapshot).
    crate::commands::sync_playback(app, engine.project().clone());
    crate::project_ipc::notify_mutation(app, engine.project());
}

/// Run a mutating engine operation; on success emit the new state.
/// Failed commands leave the engine untouched, so nothing is emitted.
fn mutate<T>(
    app: &AppHandle,
    state: &State<'_, EngineHandle>,
    op: impl FnOnce(&mut Engine) -> Result<T, cutty_engine::EngineError>,
) -> Result<T, String> {
    let mut engine = state.0.lock().expect("engine state poisoned");
    let result = op(&mut engine).map_err(|e| e.to_string())?;
    emit_state(app, &mut engine);
    Ok(result)
}

/// Fetch the current state (initial load; afterwards the frontend follows
/// `engine://project` events).
#[tauri::command]
pub fn engine_get_state(state: State<'_, EngineHandle>) -> EngineSnapshot {
    snapshot(&mut state.0.lock().expect("engine state poisoned"))
}

/// Register a media file in the project's media pool. `kind` and
/// `has_alpha` come from the probe (`probe_media`); older callers'
/// video/audio semantics are the `Video`/`Audio` kinds.
#[tauri::command]
#[allow(clippy::too_many_arguments)]
pub fn engine_add_media(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    path: String,
    duration: f64,
    has_video: bool,
    has_audio: bool,
    has_alpha: bool,
    kind: cutty_engine::MediaKind,
) -> Result<u64, String> {
    mutate(&app, &state, |e| {
        e.add_media_with_kind(path, duration, has_video, has_audio, has_alpha, kind)
            .map(|m| m.0)
    })
}

/// Remove a media file from the pool and every clip referencing it, as a
/// single undoable command.
#[tauri::command]
pub fn engine_remove_media(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    media_id: u64,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.remove_media(MediaId(media_id)))
}

/// Place a new clip on a track; returns the new clip's id.
#[tauri::command]
pub fn engine_add_clip(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    track_id: u64,
    media_id: u64,
    timeline_in: f64,
    source_in: f64,
    source_out: f64,
) -> Result<u64, String> {
    mutate(&app, &state, |e| {
        e.add_clip(
            TrackId(track_id),
            MediaId(media_id),
            timeline_in,
            source_in,
            source_out,
        )
        .map(|c| c.0)
    })
}

/// Place a new text clip. `track_id: null` = CapCut placement (topmost
/// text lane with room, else a new lane on top — one undo step). The
/// preset transform places lower-thirds etc. Returns the new clip's id.
#[tauri::command]
pub fn engine_add_text_clip(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    timeline_in: f64,
    duration: f64,
    text: TextSpec,
    transform: Option<Transform>,
    track_id: Option<u64>,
) -> Result<u64, String> {
    mutate(&app, &state, |e| {
        e.add_text_clip(
            timeline_in,
            duration,
            text,
            transform.unwrap_or_default(),
            track_id.map(TrackId),
        )
        .map(|c| c.0)
    })
}

/// Replace a text clip's payload (content and/or style). Equal payloads
/// are a no-op (no undo entry).
#[tauri::command]
pub fn engine_set_clip_text(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    text: TextSpec,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.set_clip_text(ClipId(clip_id), text))
}

/// Move a clip to a new timeline position.
#[tauri::command]
pub fn engine_move_clip(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    timeline_in: f64,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.move_clip(ClipId(clip_id), timeline_in))
}

/// Drag one edge of a clip to (approximately) `to`; returns the clamped
/// edge time actually applied.
#[tauri::command]
pub fn engine_trim_clip(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    edge: TrimEdgeArg,
    to: f64,
) -> Result<f64, String> {
    mutate(&app, &state, |e| {
        e.trim_clip(ClipId(clip_id), edge.into(), to)
    })
}

/// Split a clip at a timeline time; returns the new right half's id.
#[tauri::command]
pub fn engine_split_clip(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    at: f64,
) -> Result<u64, String> {
    mutate(&app, &state, |e| {
        e.split_clip(ClipId(clip_id), at).map(|c| c.0)
    })
}

/// Remove a clip, leaving a gap.
#[tauri::command]
pub fn engine_delete_clip(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.delete_clip(ClipId(clip_id)))
}

/// Remove a clip and shift later clips on the track left.
#[tauri::command]
pub fn engine_ripple_delete(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.ripple_delete(ClipId(clip_id)))
}

/// Set a clip's audio gain (linear; 1.0 = unity, 0.0 = silent).
#[tauri::command]
pub fn engine_set_clip_volume(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    volume: f64,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.set_clip_volume(ClipId(clip_id), volume))
}

/// Add a keyframe (or replace the one already at that time) on a clip
/// property lane. `t` is clip-relative seconds; returns the time the
/// keyframe landed on after clamping/dedup.
#[tauri::command]
pub fn engine_add_keyframe(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    prop: KeyframeProp,
    t: f64,
    value: f64,
    easing: Option<Easing>,
) -> Result<f64, String> {
    mutate(&app, &state, |e| {
        e.add_keyframe(ClipId(clip_id), prop, t, value, easing.unwrap_or_default())
    })
}

/// Move the keyframe at `from_t` to `to_t` with a new value (both axes
/// of a dot drag). Returns the applied time after neighbor clamping.
#[tauri::command]
pub fn engine_move_keyframe(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    prop: KeyframeProp,
    from_t: f64,
    to_t: f64,
    value: f64,
) -> Result<f64, String> {
    mutate(&app, &state, |e| {
        e.move_keyframe(ClipId(clip_id), prop, from_t, to_t, value)
    })
}

/// Remove the keyframe at clip-relative time `t`.
#[tauri::command]
pub fn engine_remove_keyframe(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    prop: KeyframeProp,
    t: f64,
) -> Result<(), String> {
    mutate(&app, &state, |e| {
        e.remove_keyframe(ClipId(clip_id), prop, t)
    })
}

/// Set a clip's fade-in or fade-out duration in seconds (0 removes it)
/// — sugar over the volume keyframe lane. Returns the applied duration.
#[tauri::command]
pub fn engine_set_clip_fade(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    side: FadeSide,
    duration: f64,
) -> Result<f64, String> {
    mutate(&app, &state, |e| {
        e.set_clip_fade(ClipId(clip_id), side, duration)
    })
}

/// Extract a video clip's audio onto an audio track (one undo step);
/// the video clip's own volume drops to 0. Returns the new clip's id.
#[tauri::command]
pub fn engine_extract_audio(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
) -> Result<u64, String> {
    mutate(&app, &state, |e| {
        e.extract_audio(ClipId(clip_id)).map(|c| c.0)
    })
}

/// Set a clip's 2D placement (position/scale/rotation).
#[tauri::command]
pub fn engine_set_clip_transform(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    transform: Transform,
) -> Result<(), String> {
    mutate(&app, &state, |e| {
        e.set_clip_transform(ClipId(clip_id), transform)
    })
}

/// Set a clip's opacity (0.0 transparent .. 1.0 opaque).
#[tauri::command]
pub fn engine_set_clip_opacity(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    opacity: f64,
) -> Result<(), String> {
    mutate(&app, &state, |e| {
        e.set_clip_opacity(ClipId(clip_id), opacity)
    })
}

/// Set how a clip blends with the layers below it.
#[tauri::command]
pub fn engine_set_clip_blend_mode(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    mode: BlendMode,
) -> Result<(), String> {
    mutate(&app, &state, |e| {
        e.set_clip_blend_mode(ClipId(clip_id), mode)
    })
}

/// Move a clip onto another track (and position) in one step.
#[tauri::command]
pub fn engine_move_clip_to_track(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    track_id: u64,
    timeline_in: f64,
) -> Result<(), String> {
    mutate(&app, &state, |e| {
        e.move_clip_to_track(ClipId(clip_id), TrackId(track_id), timeline_in)
    })
}

/// Insert a new empty track at panel `index` (0 = top; clamped). Returns
/// the new track's id.
#[tauri::command]
pub fn engine_add_track(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    kind: TrackKind,
    index: usize,
) -> Result<u64, String> {
    mutate(&app, &state, |e| e.add_track(kind, index).map(|t| t.0))
}

/// Remove a track with all its clips (one undo step). Rejected for the
/// last track of a kind and for locked tracks.
#[tauri::command]
pub fn engine_remove_track(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    track_id: u64,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.remove_track(TrackId(track_id)))
}

/// Move a track to another panel position (restacks the composite).
#[tauri::command]
pub fn engine_move_track(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    track_id: u64,
    to: usize,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.move_track(TrackId(track_id), to))
}

/// Flip a per-track flag: "locked" | "muted" | "hidden".
#[tauri::command]
pub fn engine_set_track_flag(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    track_id: u64,
    flag: TrackFlag,
    value: bool,
) -> Result<(), String> {
    mutate(&app, &state, |e| {
        e.set_track_flag(TrackId(track_id), flag, value)
    })
}

/// Wire form of a transition assignment.
#[derive(Clone, serde::Deserialize)]
#[serde(rename_all = "camelCase")]
pub struct TransitionArg {
    pub kind: String,
    pub duration: f64,
}

/// Set, replace, or remove (`transition: null`) the transition at a
/// clip's out cut. Returns the stored duration after clamping (`null`
/// when removing).
#[tauri::command]
pub fn engine_set_transition(
    app: AppHandle,
    state: State<'_, EngineHandle>,
    clip_id: u64,
    transition: Option<TransitionArg>,
) -> Result<Option<f64>, String> {
    mutate(&app, &state, |e| {
        e.set_transition(
            ClipId(clip_id),
            transition.map(|t| Transition {
                kind: t.kind,
                duration: t.duration,
            }),
        )
    })
}

/// Undo the most recent command; `false` when there was nothing to undo.
#[tauri::command]
pub fn engine_undo(app: AppHandle, state: State<'_, EngineHandle>) -> Result<bool, String> {
    mutate(&app, &state, |e| e.undo())
}

/// Redo the most recently undone command; `false` when there was nothing
/// to redo.
#[tauri::command]
pub fn engine_redo(app: AppHandle, state: State<'_, EngineHandle>) -> Result<bool, String> {
    mutate(&app, &state, |e| e.redo())
}

/// Open a gesture transaction (mousedown of a drag).
#[tauri::command]
pub fn engine_begin_transaction(
    app: AppHandle,
    state: State<'_, EngineHandle>,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.begin_transaction())
}

/// Commit the open transaction into a single undo entry (mouseup).
#[tauri::command]
pub fn engine_commit_transaction(
    app: AppHandle,
    state: State<'_, EngineHandle>,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.commit_transaction())
}

/// Abort the open transaction, restoring pre-gesture state (Escape).
#[tauri::command]
pub fn engine_rollback_transaction(
    app: AppHandle,
    state: State<'_, EngineHandle>,
) -> Result<(), String> {
    mutate(&app, &state, |e| e.rollback_transaction())
}

/// Snap a time against clip edges and the playhead. Read-only. The UI
/// converts its pixel snap radius to seconds before calling.
#[tauri::command]
pub fn engine_snap_time(
    state: State<'_, EngineHandle>,
    t: f64,
    threshold: f64,
    playhead: Option<f64>,
    exclude: Vec<u64>,
) -> Option<f64> {
    let engine = state.0.lock().expect("engine state poisoned");
    let exclude: Vec<ClipId> = exclude.into_iter().map(ClipId).collect();
    cutty_engine::snap_time(engine.project(), t, threshold, playhead, &exclude)
}

/// Snap a clip-move gesture (both edges compete). Read-only.
#[tauri::command]
pub fn engine_snap_clip_move(
    state: State<'_, EngineHandle>,
    clip_id: u64,
    desired_in: f64,
    threshold: f64,
    playhead: Option<f64>,
) -> Result<SnappedMove, String> {
    let engine = state.0.lock().expect("engine state poisoned");
    cutty_engine::snap_clip_move(
        engine.project(),
        ClipId(clip_id),
        desired_in,
        threshold,
        playhead,
    )
    .map_err(|e| e.to_string())
}
